packages/gatsby-source-filesystem/src/create-file-node.js

```diff
@@ -1,79 +1,74 @@
-    ...parsedSlashed,
-    absolutePath: slashed,
-    // Useful for limiting graphql query with certain parent directory
-    relativeDirectory: path.relative(
-      pluginOptions.path || process.cwd(),
-      parsedSlashed.dir
-    ),
+  node = {
+    ...node,
+    ...path.parse(node.absolutePath),
+    ...stats,
+    absPath: slash(path.resolve(node.absolutePath)),
+    relativePath: slash(path.relative(node.cwd, node.absolutePath)),
   }
 
-  const stats = await fs.stat(slashedFile.absolutePath)
-  let internal
-  if (stats.isDirectory()) {
-    const contentDigest = crypto
-      .createHash(`md5`)
-      .update(
-        JSON.stringify({ stats: stats, absolutePath: slashedFile.absolutePath })
-      )
-      .digest(`hex`)
-    internal = {
-      contentDigest,
-      type: `Directory`,
-    }
-  } else {
-    const contentDigest = await md5File(slashedFile.absolutePath)
-    const mediaType = mime.getType(slashedFile.ext)
-    internal = {
-      contentDigest,
-      type: `File`,
-      mediaType: mediaType ? mediaType : `application/octet-stream`,
-    }
+  node = {
+    ...node,
+    accessTime: node.atime,
+    birthTime: node.birthtime,
+    changeTime: node.ctime,
+    extension: node.ext.slice(1).toLowerCase(),
+    internal: stats.isDirectory()
+      ? {
+          contentDigest: crypto
+            .createHash(`md5`)
+            .update(
+              JSON.stringify({
+                absolutePath: node.absolutePath,
+                stats: stats,
+              })
+            )
+            .digest(`hex`),
+          type: `Directory`,
+        }
+      : {
+          contentDigest: await md5File(node.absolutePath),
+          mediaType: mime.getType(node.ext) || `application/octet-stream`,
+          type: `File`,
+        },
+    modifiedTime: node.mtime,
+    prettySize: prettyBytes(node.size),
+    relativeDirectory: path.relative(node.cwd, node.dir),
   }
 
-  // Stringify date objects.
-  return JSON.parse(
-    JSON.stringify({
-      // Don't actually make the File id the absolute path as otherwise
-      // people will use the id for that and ids shouldn't be treated as
-      // useful information.
-      id: createNodeId(pathToFile),
-      children: [],
-      parent: `___SOURCE___`,
-      internal,
-      sourceInstanceName: pluginOptions.name || `__PROGRAMATTIC__`,
-      absolutePath: slashedFile.absolutePath,
-      relativePath: slash(
-        path.relative(
-          pluginOptions.path || process.cwd(),
-          slashedFile.absolutePath
-        )
-      ),
-      extension: slashedFile.ext.slice(1).toLowerCase(),
-      size: stats.size,
-      prettySize: prettyBytes(stats.size),
-      modifiedTime: stats.mtime,
-      accessTime: stats.atime,
-      changeTime: stats.ctime,
-      birthTime: stats.birthtime,
-      ...slashedFile,
-      ...stats,
-    })
-  )
+  node = {
+    ...node,
+    allFile: {
+      edges: stats.isDirectory()
+        ? await fs
+            .readdir(node.absolutePath)
```

Comment:
I wonder how expensive this is? If it's really cheap, it's fine but we already have read all this data so it could get expensive to double-read information all the way through a deep file tree.

Reply:
There very well may be a better way of accomplishing this, but I'm not sure what it would be.